### ISSUE TITLE
Regression introduced in z3-4.8.10 - unsat to unknown

### ISSUE THREAD

**Issue by radumereuta** [Z3Prover/z3#5243]
```
(define-sort IMap () (Array Int Int))
(declare-fun isValidBranch (IMap Int) Bool)
(declare-fun isValidZeros (IMap Int) Bool)
(declare-fun |K_138| () Int)
(declare-fun |TREE_HEIGHT_137| () Int)
(declare-fun |M_139| () Int)
(declare-fun |Branch_135| () IMap)
(declare-fun |Zeros_136| () IMap)
(assert (and
  (and
    (= (>= |K_138| |TREE_HEIGHT_137|) true)
    (= (<= |K_138| |TREE_HEIGHT_137|) true)
    (= (< |M_139| (^ 2 |TREE_HEIGHT_137|)) true)
    (= (isValidBranch |Branch_135| |M_139|) true)
    (= (isValidZeros |Zeros_136| |M_139|) true)
    (= (>= |K_138| 0) true)
    (= (>= |M_139| 0) true)
    (= (>= |TREE_HEIGHT_137| 0) true)
  )
  (not (and
    (= (<= (+ |M_139| 1) (^ 2 |K_138|)) true)
  ))
))

(check-sat)
```
z3-4.8.7 -> unsat
z3-4.8.8 -> unsat
z3-4.8.9 -> unsat
z3-4.8.10 -> unknown
z3-master -> unknown (0c6722f48b2814e44ed98ea33b2a356579fd90f4)
Tested with Ubuntu 21.04, 64 bit. Built from sources.

**Comment by NikolajBjorner** [comment]
the operator ^ is non-standard (not part of SMTLIB2 definition).
The implementation prior to 4.8.10 assumed that when given two integers the result would be integer.
This assumption is wrong when the exponent is negative, so generally it returns a Real (while your context restricts the exponents to be non-negative it requires context for type inference to know this).
Unfortunately, when we don't know anything more about exponentiation and only know that the result is Real, the constraints are not shown unsat using just integer arithmetic.
You can "patch" this by inserting (to_int (^... ...)) in the places you use power. In this way you are saying you only care about integer interpretations of ^. You will get the old behavior then.

**Comment by radumereuta** [comment]
Thanks for the prompt reply.
This workaround works for the master build but does not work with z3-4.8.10.
If it's going to be included in z3-4.8.11, then I can work with that.
